pkg/scanner/scan.go

```diff
@@ -34,35 +33,74 @@ func AddAcceptanceExtensions(keys []string) {
 	}
 }
 
-func ScanImage(ctx context.Context, imageName, filePath string, dockerOption deckodertypes.DockerOption) (assessments []*types.Assessment, err error) {
-	var ext extractor.Extractor
-	var cleanup func()
-	if imageName != "" {
-		ext, cleanup, err = docker.NewDockerExtractor(ctx, imageName, dockerOption)
-		if err != nil {
-			return nil, fmt.Errorf("%v. %w", err, types.ErrorCreateDockerExtractor)
-		}
-	} else if filePath != "" {
-		ext, cleanup, err = docker.NewDockerArchiveExtractor(ctx, filePath, dockerOption)
-		if err != nil {
-			return nil, fmt.Errorf("%v. %w", err, types.ErrorCreateDockerExtractor)
-		}
+func ScanImage(ctx context.Context, cfg config.Config) ([]*types.Assessment, error) {
+	registryOptions := image.RegistryOptions{
+		InsecureSkipTLSVerify: config.Conf.Insecure,
+		Credentials: []image.RegistryCredentials{
+			{
+				Username: config.Conf.Username,
+				Password: config.Conf.Password,
+				Token:    config.Conf.Token,
+			},
+		},
+		InsecureUseHTTP: config.Conf.NonSSL,
+		Platform:        config.Conf.Platform,
+	}
+
+	var userInput string
+	if cfg.FilePath != "" {
+		userInput = cfg.FilePath
+	} else if cfg.ImageName != "" {
+		userInput = setImageSource(cfg.LocalImage, cfg.ImageName)
 	} else {
 		return nil, types.ErrSetImageOrFile
 	}
-	defer cleanup()
-	ac := analyzer.New(ext)
-	var files deckodertypes.FileMap
+
+	opts := stereoscope.WithRegistryOptions(registryOptions)
+	img, err := stereoscope.GetImage(ctx, userInput, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get image from user input=%s: %w", userInput, err)
+	}
+
+	files := make(map[string]types.FileData)
 	filterFunc := createPathPermissionFilterFunc(assessor.LoadRequiredFiles(), assessor.LoadRequiredExtensions(), assessor.LoadRequiredPermissions())
-	if files, err = ac.Analyze(ctx, filterFunc); err != nil {
-		return nil, fmt.Errorf("%v. %w", err, types.ErrorAnalyze)
+
+	files["/config"] = types.FileData{
+		Body: img.Metadata.RawConfig,
+	}
+	refs := img.SquashedTree().AllFiles()
+	for i := range refs {
+		entry, err := img.FileCatalog.Get(refs[i])
+		if err != nil {
+			return nil, fmt.Errorf("failed to get entry from file catalog reference=%+v: %w", refs[i], err)
+		}
+		fileMode := entry.FileInfo.Mode()
+		ok, err := filterFunc(entry.Path, fileMode)
+		if err != nil {
+			return nil, fmt.Errorf("failed to run filter function on file=%s: %w", entry.RealPath, err)
+		}
+		if !ok {
+			continue
+		}
+
+		contentReader, err := img.OpenPathFromSquash(entry.RealPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open file=%s: %w", entry.RealPath, err)
+		}
+		content, err := io.ReadAll(contentReader)
```

Comment:
This will explode the memory to keep the whole image in memory. Can we save the content reader and access it when needed?

Reply:
It is working the same way as before. It doesn't read all of the files, just the necessary ones based on the filters. But I can update it.

Reply:
I've updated.

Reply:
Great! Somebody should close it?